services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/batchmanager/ReceiveQueueBuffer.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.sqs.internal.batchmanager;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReentrantLock;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.services.sqs.SqsAsyncClient;
+import software.amazon.awssdk.services.sqs.model.Message;
+import software.amazon.awssdk.services.sqs.model.ReceiveMessageResponse;
+
+@SdkInternalApi
+public class ReceiveQueueBuffer {
+
+    private final ScheduledExecutorService executor;
+    private final SqsAsyncClient sqsClient;
+    private final ResponseBatchConfiguration config;
+    private final String queueUrl;
+    private final  QueueAttributesManager queueAttributesManager;
+
+    private final Queue<AsyncReceiveMessageBatch> finishedTasks = new ConcurrentLinkedQueue<>();
+    private final Queue<ReceiveMessageCompletableFuture> futures = new ConcurrentLinkedQueue<>();
+
+    private final AtomicInteger inflightReceiveMessageBatches = new AtomicInteger(0);
+    private final AtomicBoolean shutDown = new AtomicBoolean(false);
+
+    private final ReentrantLock lock = new ReentrantLock();
+
+    public ReceiveQueueBuffer(ScheduledExecutorService executor, SqsAsyncClient sqsClient,
+                              ResponseBatchConfiguration config, String queueUrl, QueueAttributesManager queueAttributesManager) {
+        this.executor = executor;
+        this.sqsClient = sqsClient;
+        this.config = config;
+        this.queueUrl = queueUrl;
+        this.queueAttributesManager = queueAttributesManager;
+    }
+
+    public void receiveMessage(ReceiveMessageCompletableFuture receiveMessageFuture) {
+        futures.add(receiveMessageFuture);
+        satisfyFuturesFromBuffer();
```

Comment:
It seems we are triggering this delivering events until it's empty kind of thing whenever there's a request demand. Is this any reason we can't just pull from `finishedTasks` and complete this future?

Reply:
Yeap , we want to make sure the older request which were waiting for the recieverMessage call from SQS gets processed in first , else they get starved and will return empty messages (Same behavior as V1)